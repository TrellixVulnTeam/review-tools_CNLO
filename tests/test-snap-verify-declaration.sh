#!/bin/sh
set -e

help() {
    cat <<EOM
$ $(basename "$0")
EOM
}

if [ "$1" = "help" ] || [ "$1" = "--help" ] || [ "$1" = "-h" ]; then
    help
    exit
fi

tmpdir=$(mktemp -d)
tmp="$tmpdir/out"
orig_dir=$(pwd)

run() {
    echo "Running: snap-verify-declaration $*" | tee -a "$tmp"
    run_no_header "$@" 2>&1 | tee -a "$tmp"
    echo "" | tee -a "$tmp"
}

run_no_header() {
    PYTHONPATH="$orig_dir" "$orig_dir"/bin/snap-verify-declaration "$@"
}

comment() {
    echo "$@" | tee -a "$tmp"
}

origdir="$(pwd)"
cd "$tmpdir"

# -h
comment "= Test -h ="
run "-h"

# plugs
cat > "./plugs" <<EOM
{
  "home": {
    "allow-auto-connection": "true",
    "allow-connection": "true",
    "allow-installation": "true"
  }
}
EOM
run --plugs="./plugs"
run --json --plugs="./plugs"

# slots
cat > "./slots" <<EOM
{
  "mir": {
    "allow-auto-connection": "true",
    "allow-connection": "true",
    "allow-installation": "true"
  }
}
EOM
run --slots="./slots"
run --json --slots="./slots"

# now do both
run --plugs="./plugs" --slots="./slots"
run --json --plugs="./plugs" --slots="./slots"

# empty is ok
cat > "./plugs" <<EOM
{}
EOM
run --plugs="./plugs"
run --json --plugs="./plugs"

# malformed
cat > "./plugs" <<EOM
{
EOM
run --plugs="./plugs"
run --json --plugs="./plugs"

# nonexistent interface
cat > "./plugs" <<EOM
{
  "nonexistent": {
    "allow-auto-connection": "true"
  }
}
EOM
run --plugs="./plugs"
run --json --plugs="./plugs"

# invalid (plug-publisher-id with plugs (should be slot-publisher-id))
cat > "./plugs" <<EOM
{
  "mir": {
    "allow-auto-connection": {
      "plug-publisher-id": [
        "aaaaaaaaaabbbbbbbbbbcccccccccc32"
      ]
    }
  }
}
EOM
run --plugs="./plugs"
run --json --plugs="./plugs"


cd "$origdir"
echo
echo "Checking for differences in output..."
diff -Naur ./tests/test-snap-verify-declaration.sh.expected "$tmp" || {
   echo "Found unexpected differences between './tests/test-snap-verify-declaration.sh.expected' and '$tmp'"
   exit 1
}

rm -f "$tmp" "$tmpdir"/plugs* "$tmpdir"/slots*
rmdir "$tmpdir"
echo "Done"
