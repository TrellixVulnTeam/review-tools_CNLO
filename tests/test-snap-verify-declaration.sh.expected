= Test -h =
Running: snap-verify-declaration -h
usage: snap-verify-declaration [-h] [--json] [--plugs PLUGS] [--slots SLOTS]

Check a snap declaration for errors

optional arguments:
  -h, --help     show this help message and exit
  --json         print json output
  --plugs PLUGS  file specifying snap declaration for plugs
  --slots SLOTS  file specifying snap declaration for slots

RETURN CODES
  0     found no errors or warnings
  1     checks not run due to fatal error
  2     found only errors or errors and warnings
  3     found only warnings

Running: snap-verify-declaration --plugs=./plugs
{'error': {},
 'info': {'snap-declaration-verify_v2:valid_plugs:home:allow-auto-connection': {'manual_review': False,
                                                                                'text': 'OK'},
          'snap-declaration-verify_v2:valid_plugs:home:allow-connection': {'manual_review': False,
                                                                           'text': 'OK'},
          'snap-declaration-verify_v2:valid_plugs:home:allow-installation': {'manual_review': False,
                                                                             'text': 'OK'}},
 'warn': {}}

Running: snap-verify-declaration --json --plugs=./plugs
{
  "error": {},
  "info": {
    "snap-declaration-verify_v2:valid_plugs:home:allow-auto-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_plugs:home:allow-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_plugs:home:allow-installation": {
      "manual_review": false,
      "text": "OK"
    }
  },
  "warn": {}
}

Running: snap-verify-declaration --slots=./slots
{'error': {},
 'info': {'snap-declaration-verify_v2:valid_slots:mir:allow-auto-connection': {'manual_review': False,
                                                                               'text': 'OK'},
          'snap-declaration-verify_v2:valid_slots:mir:allow-connection': {'manual_review': False,
                                                                          'text': 'OK'},
          'snap-declaration-verify_v2:valid_slots:mir:allow-installation': {'manual_review': False,
                                                                            'text': 'OK'}},
 'warn': {}}

Running: snap-verify-declaration --json --slots=./slots
{
  "error": {},
  "info": {
    "snap-declaration-verify_v2:valid_slots:mir:allow-auto-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_slots:mir:allow-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_slots:mir:allow-installation": {
      "manual_review": false,
      "text": "OK"
    }
  },
  "warn": {}
}

Running: snap-verify-declaration --plugs=./plugs --slots=./slots
{'error': {},
 'info': {'snap-declaration-verify_v2:valid_plugs:home:allow-auto-connection': {'manual_review': False,
                                                                                'text': 'OK'},
          'snap-declaration-verify_v2:valid_plugs:home:allow-connection': {'manual_review': False,
                                                                           'text': 'OK'},
          'snap-declaration-verify_v2:valid_plugs:home:allow-installation': {'manual_review': False,
                                                                             'text': 'OK'},
          'snap-declaration-verify_v2:valid_slots:mir:allow-auto-connection': {'manual_review': False,
                                                                               'text': 'OK'},
          'snap-declaration-verify_v2:valid_slots:mir:allow-connection': {'manual_review': False,
                                                                          'text': 'OK'},
          'snap-declaration-verify_v2:valid_slots:mir:allow-installation': {'manual_review': False,
                                                                            'text': 'OK'}},
 'warn': {}}

Running: snap-verify-declaration --json --plugs=./plugs --slots=./slots
{
  "error": {},
  "info": {
    "snap-declaration-verify_v2:valid_plugs:home:allow-auto-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_plugs:home:allow-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_plugs:home:allow-installation": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_slots:mir:allow-auto-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_slots:mir:allow-connection": {
      "manual_review": false,
      "text": "OK"
    },
    "snap-declaration-verify_v2:valid_slots:mir:allow-installation": {
      "manual_review": false,
      "text": "OK"
    }
  },
  "warn": {}
}

Running: snap-verify-declaration --plugs=./plugs
{'error': {}, 'info': {}, 'warn': {}}

Running: snap-verify-declaration --json --plugs=./plugs
{
  "error": {},
  "info": {},
  "warn": {}
}

Running: snap-verify-declaration --plugs=./plugs
ERROR: Could not read plugs: Expecting property name enclosed in double quotes: line 2 column 1 (char 2)

Running: snap-verify-declaration --json --plugs=./plugs
{
  "runtime-errors": {
    "error": {
      "msg": {
        "manual_review": true,
        "text": "Could not read plugs: Expecting property name enclosed in double quotes: line 2 column 1 (char 2)"
      }
    },
    "info": {},
    "warn": {}
  }
}

Running: snap-verify-declaration --plugs=./plugs
{'error': {'snap-declaration-verify_v2:valid_interface:nonexistent': {'manual_review': False,
                                                                      'text': 'interface '
                                                                              "'nonexistent' "
                                                                              'not '
                                                                              'found '
                                                                              'in '
                                                                              'base '
                                                                              'declaration'}},
 'info': {'snap-declaration-verify_v2:valid_plugs:nonexistent:allow-auto-connection': {'manual_review': False,
                                                                                       'text': 'OK'}},
 'warn': {}}

Running: snap-verify-declaration --json --plugs=./plugs
{
  "error": {
    "snap-declaration-verify_v2:valid_interface:nonexistent": {
      "manual_review": false,
      "text": "interface 'nonexistent' not found in base declaration"
    }
  },
  "info": {
    "snap-declaration-verify_v2:valid_plugs:nonexistent:allow-auto-connection": {
      "manual_review": false,
      "text": "OK"
    }
  },
  "warn": {}
}

Running: snap-verify-declaration --plugs=./plugs
{'error': {'snap-declaration-verify_v2:valid_plugs:mir:allow-auto-connection_plug-publisher-id': {'manual_review': False,
                                                                                                  'text': 'declaration '
                                                                                                          'malformed '
                                                                                                          '(unknown '
                                                                                                          'constraint '
                                                                                                          'key '
                                                                                                          "'plug-publisher-id')"}},
 'info': {},
 'warn': {}}

Running: snap-verify-declaration --json --plugs=./plugs
{
  "error": {
    "snap-declaration-verify_v2:valid_plugs:mir:allow-auto-connection_plug-publisher-id": {
      "manual_review": false,
      "text": "declaration malformed (unknown constraint key 'plug-publisher-id')"
    }
  },
  "info": {},
  "warn": {}
}

